### DISCUSSION TITLE
Shifted labels after auto segmentation caused by empty segments

### DISCUSSION THREAD

**Discussion by marianabb** [Project-MONAI/MONAILabel#809] [Q&A] [answered]
Hello,

I am testing the 'the radiology/segmentation' app with my own pretrained model and my local data for organ segmentation on abdominal CT. I am using MONAILabel version 0.4.0rc3 and 3D Slicer 5.1.0 r30918.

I've noticed that if any of the labels are empty, for example, in case an organ happens to be out of frame in the particular image, then the existing segmentations are shifted when the result is loaded to Slicer after clicking 'Run' in the Auto segmentation function.

Looking a bit into it, the result json file created in 'InferTask' includes the correct labels in "label_names", and the temporary file created during inference also contains the correct label numbers, although this seems to not be enough for Slicer to load the segmentation correctly (I found this related discussion: https://discourse.slicer.org/t/empty-segments-cause-shift-in-segment-index-when-reloaded/11212).

Here is an example with a network pretrained for organ segmentation where the bladder is only included in some images:

* Correct result when the bladder is included:
![bladder](https://user-images.githubusercontent.com/506780/170722158-3b022435-9b65-4efe-9d36-89abda27151d.png)

and the output in Slicer's console:
`Result Params for Segmentation: {'label_names': {'liver': 1, 'bladder': 2, 'lungs': 3, 'kidneys': 4, 'bone': 5, 'brain': 6}, 'latencies': {'pre': 3.7, 'infer': 3.15, 'invert': 0.0, 'post': 2.15, 'write': 1.31, 'total': 10.31}}`
`Update Segmentation Mask using Labels: ['liver', 'bladder', 'lungs', 'kidneys', 'bone', 'brain']`
`Import only Freezed label: None`
`Setting new segmentation with id: jake => jake`
`Setting new segmentation with id: Peach => Peach`
`Setting new segmentation with id: Brain => Brain`
`Setting new segmentation with id: Ventricles => Ventricles`
`Setting new segmentation with id: Vessels => Vessels`

* Incorrect result when the bladder is not included (shifted colors and label names):
![no_bladder](https://user-images.githubusercontent.com/506780/170722650-ae6fc4a2-bcd9-4c53-af40-5f9aa43d4653.png)

the output in Slicer's console shows all the labels, but only creates four segmentation ids:
`Result Params for Segmentation: {'label_names': {'liver': 1, 'bladder': 2, 'lungs': 3, 'kidneys': 4, 'bone': 5, 'brain': 6}, 'latencies': {'pre': 2.31, 'infer': 3.16, 'invert': 0.0, 'post': 2.19, 'write': 0.25, 'total': 7.91}}`
`Update Segmentation Mask using Labels: ['liver', 'bladder', 'lungs', 'kidneys', 'bone', 'brain']`
`Import only Freezed label: None`
`Setting new segmentation with id: jake => jake`
`Setting new segmentation with id: Brain => Brain`
`Setting new segmentation with id: Ventricles => Ventricles`
`Setting new segmentation with id: Vessels => Vessels`

This may not be a big deal in cases like the multiorgan segmentation example (although it will probably fail if a patient happens to have only one kidney, for example), but it is very bad for other cases such as vertebrae segmentation when all the possible vertebrae are usually not included in many of the images.

Any ideas of what would be the easiest way to solve this?

Thanks!

**Comment by diazandr3s** [comment, marked as answer]
Hi @marianabb,

Thanks for the detailed comment.
We've been discussing how to sort of standardizing the color for the different segments. Here @lassoan [opened an issue in the NGFF repo](https://github.com/ome/ngff/issues/79) to see the interest of the community.

For the MONAI Label module we use these colors to represent the different organs: https://github.com/Project-MONAI/MONAILabel/blob/main/plugins/slicer/MONAILabel/MONAILabelLib/labelcolors.py#L1

If an organ is not on the list, it takes a random color: https://github.com/Project-MONAI/MONAILabel/blob/main/plugins/slicer/MONAILabel/MONAILabel.py#L1621  @SachidanandAlle please keep me honest here

My suggestion for having the same color is to follow the same organ names as in the list. For instance, I'd call [urinary bladder](https://github.com/Project-MONAI/MONAILabel/blob/main/plugins/slicer/MONAILabel/MONAILabelLib/labelcolors.py#L228) to what you called bladder, etc

Hope this helps

**Reply by marianabb** [reply]
Hi @diazandr3s, thank you for the reply!

I think the problem is not the colors but the fact that the labels are shifted upwards when some of them are empty. In the example, see how the lungs, which are supposed to be yellow "became" bladder, the kidneys are now lungs, and bone is now kidneys, leaving the last two labels empty. This happens also if I use the predefined names as you suggest:

![no_bladder2](https://user-images.githubusercontent.com/506780/170964322-26170d9c-5b62-4fc6-80f0-11014a74aeef.png)

The colors are now a bit harder to differentiate, but notice how the bone label is now empty and its contents have been shifted to "right kidney", the same happened to the other labels that were under "urinary bladder" which happened to be empty in this case.

If I now use Slicer to save this segmentation to a file the label numbers are also shifted, so they don't match my Task-defined labels anymore.

**Reply by lassoan** [reply]
Using a color table (that specifies mapping between segments and label values) for both importing and exporting labelmap volume should fix the above mentioned issues. See code snippets here: https://slicer.readthedocs.io/en/latest/developer_guide/script_repository.html#import-export-labelmap-node-using-custom-label-value-mapping

**Reply by marianabb** [reply]
I managed to solve the problem with a few modifications to the function `updateSegmentationMask` in the Slicer module. The main change was to add an explicit list of segment IDs as an argument to the function `ImportLabelmapToSegmentationNode`, which then makes it unnecessary to differentiate between 'existing segments' and 'added segments' in the rest of the code. Here is the new function:

```python
def updateSegmentationMask(self, in_file, labels, sliceIndex=None, freeze=None):
  # TODO:: Add ROI Node (for Bounding Box if provided in the result)
  start = time.time()
  logging.debug(f"Update Segmentation Mask from: {in_file}")
  if in_file and not os.path.exists(in_file):
      return False

  segmentationNode = self._segmentNode
  segmentation = segmentationNode.GetSegmentation()

  if in_file is None:
      for label in labels:
          if not segmentation.GetSegmentIdBySegmentName(label):
              segmentation.AddEmptySegment(label, label, self.getLabelColor(label))
      return True

  if in_file.endswith(".seg.nrrd") and self.file_ext == ".seg.nrrd":
      source_node = slicer.modules.segmentations.logic().LoadSegmentationFromFile(in_file, False)
      destination_node = segmentationNode
      destination_segmentations = destination_node.GetSegmentation()
      source_segmentations = source_node.GetSegmentation()

      destination_segmentations.DeepCopy(source_segmentations)

      if self._volumeNode:
          destination_node.SetReferenceImageGeometryParameterFromVolumeNode(self._volumeNode)

      slicer.mrmlScene.RemoveNode(source_node)
  else:
      labels = [label for label in labels if label != "background"]
      print(f"Update Segmentation Mask using Labels: {labels}")

      # segmentId, segment = self.currentSegment()
      labelImage = sitk.ReadImage(in_file)
      labelmapVolumeNode = sitkUtils.PushVolumeToSlicer(labelImage, None, className="vtkMRMLLabelMapVolumeNode")

      existing_label_ids = {}
      for label in labels:
          id = segmentation.GetSegmentIdBySegmentName(label)
          if id:
              existing_label_ids[label] = id

      freeze = [freeze] if freeze and isinstance(freeze, str) else freeze
      print(f"Import only Freezed label: {freeze}")

      numberOfExistingSegments = segmentation.GetNumberOfSegments()

      # Added list of segments to load
      segmentIds = vtk.vtkStringArray() 
      for label in labels:
          segmentIds.InsertNextValue(label)
      slicer.modules.segmentations.logic().ImportLabelmapToSegmentationNode(labelmapVolumeNode, segmentationNode, segmentIds) # Additional arg segmentIds
      slicer.mrmlScene.RemoveNode(labelmapVolumeNode)

      # numberOfAddedSegments = segmentation.GetNumberOfSegments() - numberOfExistingSegments
      # logging.debug(f"Adding {numberOfAddedSegments} segments")

      # addedSegmentIds = [
      #   segmentation.GetNthSegmentID(numberOfExistingSegments + i) for i in range(numberOfAddedSegments)
      # ]

      for i, segmentId in enumerate(labels): # Iterate over all the labels
          segment = segmentation.GetSegment(segmentId)
          print(f"Setting new segmentation {i} with id: {segmentId} => {segment.GetName()}")

          label = labels[i] if i < len(labels) else f"unknown {i}"
          # segment.SetName(label)
          # segment.SetColor(self.getLabelColor(label))

          if freeze and label not in freeze:
              print(f"Discard label update for: {label}")
          elif label in existing_label_ids:
              self.ui.embeddedSegmentEditorWidget.setSegmentationNode(segmentationNode)
              self.ui.embeddedSegmentEditorWidget.setMasterVolumeNode(self._volumeNode)
              self.ui.embeddedSegmentEditorWidget.setCurrentSegmentID(existing_label_ids[label])

              effect = self.ui.embeddedSegmentEditorWidget.effectByName("Logical operators")
              labelmap = slicer.vtkOrientedImageData()
              segmentationNode.GetBinaryLabelmapRepresentation(segmentId, labelmap)

              if sliceIndex:
                  selectedSegmentLabelmap = effect.selectedSegmentLabelmap()
                  dims = selectedSegmentLabelmap.GetDimensions()
                  count = 0
                  for x in range(dims[0]):
                      for y in range(dims[1]):
                          if selectedSegmentLabelmap.GetScalarComponentAsDouble(x, y, sliceIndex, 0):
                              count = count + 1
                          selectedSegmentLabelmap.SetScalarComponentFromDouble(x, y, sliceIndex, 0, 0)

                  logging.debug(f"Total Non Zero: {count}")

                  # Clear the Slice
                  if count:
                      effect.modifySelectedSegmentByLabelmap(
                          selectedSegmentLabelmap, slicer.qSlicerSegmentEditorAbstractEffect.ModificationModeSet
                      )

                  # Union label map
                  effect.modifySelectedSegmentByLabelmap(
                      labelmap, slicer.qSlicerSegmentEditorAbstractEffect.ModificationModeAdd
                  )
              else:
                  # adding bypass masking to not overwrite other layers,
                  # needed for preserving scribbles during updates
                  # help from: https://github.com/Slicer/Slicer/blob/master
                  #            /Modules/Loadable/Segmentations/EditorEffects/Python/SegmentEditorLogicalEffect.py
                  bypassMask = True
                  effect.modifySelectedSegmentByLabelmap(
                      labelmap, slicer.qSlicerSegmentEditorAbstractEffect.ModificationModeSet, bypassMask
                  )

          #segmentationNode.RemoveSegment(segmentId) # Commented since there are no double labels anymore

  self.showSegmentationsIn3D()
  logging.info(f"Time consumed by updateSegmentationMask: {time.time() - start:3.1f}")
  return True
```

Including a color table as @lassoan suggested would probably solve the issue also, but this option seemed to require less changes.

I am very new to 3D Slicer modules and also not completely familiar with all the MONAILabel code, so I hope these changes are not breaking something else, especially since this function is being called from other places in the module. Please let me know if you think this is the case!

**Reply by SachidanandAlle** [reply]
Can you please raise a PR.. to support this fix for everyone..

**Reply by marianabb** [reply]
Here is the PR: https://github.com/Project-MONAI/MONAILabel/pull/817
